diff --git a/java/util/src/main/java/com/google/protobuf/util/JsonFormat.java b/java/util/src/main/java/com/google/protobuf/util/JsonFormat.java
index 509258813..8e841e4c9 100644
--- a/java/util/src/main/java/com/google/protobuf/util/JsonFormat.java
+++ b/java/util/src/main/java/com/google/protobuf/util/JsonFormat.java
@@ -207,6 +207,21 @@ public class JsonFormat {
           sortingMapKeys);
     }
 
+    public Printer alwaysPrintFieldsWithNoPresence() {
+      if (shouldPrintDefaults != ShouldPrintDefaults.ONLY_IF_PRESENT) {
+        throw new IllegalStateException("Only one of the JsonFormat defaults options can be set.");
+      }
+      return new Printer(
+          registry,
+          oldRegistry,
+          ShouldPrintDefaults.ALWAYS_PRINT_WITHOUT_PRESENCE_FIELDS,
+          ImmutableSet.of(),
+          preservingProtoFieldNames,
+          omittingInsignificantWhitespace,
+          printingEnumsAsInts,
+          sortingMapKeys);
+    }
+
     /**
      * Creates a new {@link Printer} that will print any field that does not support presence even
      * if it would not otherwise be printed (empty repeated fields, empty map fields, and implicit
